Enum entities for login and purchase statuses & common refactoring (fixes #36)

lmalakhova opened the pull request:
- Сreate enum entities for login and purchase statuses;
- Custom Assertions common refactoring;
- Delete login and purchase status fields from Product entities;
- Amazon Tests Assertions common refactoring.

lmalakhova committed:
Added the following:
- Сreate enum entities for login and purchase statuses;
- Custom Assertions common refactoring;
- Delete login and purchase status fields from Product entities;
- Amazon Tests Assertions common refactoring.

sskorol commented on src/test/java/io/github/sskorol/testcases/AmazonTests.java:
```diff
@@ -33,7 +35,7 @@ public void shouldSearchForParfume(final Parfume parfume, final Account account)
                 .login(account.getUsername(), account.getPassword());
 
-        customAssertThat(account)
+        customAssertThat(SUCCESS)
```
you've mixed up expected and actual results... asserts accept actual first, and then verify it against expected...

sskorol commented on src/main/java/io/github/sskorol/model/LoginStatus.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum LoginStatus {
```
you may want to run quality checks on the entire project and fix violations

sskorol commented on src/main/java/io/github/sskorol/model/Product.java:
```diff
@@ -10,11 +10,9 @@
-    private final String purchaseStatus;
 
-    public Product(final String name, final String generalCategory, final String purchaseStatus) {
+    public Product(final String name, final String generalCategory) {
```
what about adding a `subCategory` field (as discussed on lesson)?

lmalakhova committed:
Added the following:
- Created enum entities for login and purchase statuses;
- SubCategory field to Product class;
- Custom Assertions common refactoring;
- Delete login and purchase status fields from Product entities;
- Amazon Tests Assertions common refactoring.

mkgerasimenko commented on src/main/java/io/github/sskorol/model/PurchaseStatus.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum PurchaseStatus {
```
Maybe we can create one enum for purchase and login status?

mkgerasimenko commented on src/test/java/io/github/sskorol/testcases/AmazonTests.java:
```diff
@@ -110,8 +109,8 @@ public void shouldSearchForShoes(final Shoes shoes, final Account account) {
-        customAssertThat(shoes)
-                .hasPurchaseStatus(at(ProductPage.class).getPurchaseStatus());
+        customAssertThat(at(ProductPage.class))
+                .hasPurchaseStatus(PurchaseStatus.SUCCESSFUL.getPurchaseStatus());
```
if we use one enum for statuses we will use the more obvious name for getter and static import
`LOGIN_SUCCESSFUL`/`PURCHASE_FAILED`.
this line would look so:
`.hasPurchaseStatus(PURCHASE_SUCCESSFUL.getStatus());`

mkgerasimenko commented on src/main/java/io/github/sskorol/assertions/CustomAssertions.java:
```diff
@@ -12,11 +12,11 @@ private CustomAssertions() {
 
-    public static LoginPageAssert customAssertThat(final Account account) {
-        return new ModelSoftAssert().assertThat(LoginPageAssert.class, Account.class, account);
+    public static LoginAssert customAssertThat(final LoginPage loginPage) {
```
I think you will change your custom assert name to more obvious. For example:
`LoginPageAssert`

mkgerasimenko commented on src/main/java/io/github/sskorol/assertions/CustomAssertions.java:
```diff
@@ -12,11 +12,11 @@ private CustomAssertions() {
 
-    public static ProductPageAssert customAssertThat(final Product product) {
-        return new ModelSoftAssert().assertThat(ProductPageAssert.class, Product.class, product);
+    public static PurchaseAssert customAssertThat(final ProductPage productPage) {
```
same as for `LoginAssert`

lmalakhova committed:
Added the following:
- Replace two Enam classes for login and purchase statuses by OperationStatus Enam;
- Rename Assert Classes;
- Amazon Tests Assertions common refactoring.

emilshina commented on src/main/java/io/github/sskorol/assertions/LoginPageAssert.java:
```diff
@@ -1,29 +1,30 @@
 
-    public LoginPageAssert(final Account account) {
-        super(account, LoginPageAssert.class);
+    public LoginPageAssert(final LoginPage actual) {
```
Rename "actual" to "loginPage" for avoiding misunderstanding. Because actual - it's protected final ACTUAL actual from AbstractAssert class.

sskorol commented on src/main/java/io/github/sskorol/model/OperationStatus.java:
```diff
@@ -0,0 +1,20 @@
+    LOGIN_SUCCESSFUL("Authorized successfully"),
+    LOGIN_FAILED("Cannot find an account with such credentials");
+
+    private final String operationStatus;
```
It makes no sense to call your instance var the same as actual class, because when you call getter, it looks a bit strange:
```java
LOGIN_FAILED.getOperationStatus();
```
Would be much more obvious to use `name` or smth similar.

sskorol commented on src/main/java/io/github/sskorol/assertions/ProductPageAssert.java:
```diff
@@ -1,29 +1,30 @@
-    @Step("Verify that the purchase status is equal to \"{purchaseStatus}\".")
-    public ProductPageAssert hasPurchaseStatus(final String purchaseStatus) {
+    @Step("Verify that the login status is equal to \"{status}\".")
+    public ProductPageAssert hasPurchaseStatus(final String status) {
```
why don't you pass enum as a parameter?

emilshina commented on src/main/java/io/github/sskorol/assertions/LoginPageAssert.java:
```diff
@@ -1,29 +1,30 @@
-        if (!Objects.areEqual(status, loginStatus)) {
-            failWithMessage(ERROR_MESSAGE_TEMPLATE, "Status name", actual, loginStatus, status);
+        if (!Objects.areEqual(loginStatus, value)) {
+            failWithMessage(ERROR_MESSAGE_TEMPLATE, "Status name", actual, loginStatus);
```
Need to pass 4 strings into ERROR_MESSAGE_TEMPLATE. You missed the expected status ("value").

emilshina commented on src/main/java/io/github/sskorol/assertions/LoginPageAssert.java:
```diff
@@ -1,29 +1,30 @@
-    public LoginPageAssert hasLoginStatus(final String loginStatus) {
+    @Step("Verify that the login status is equal to \"{value}\".")
+
+    public LoginPageAssert hasLoginStatus(final String value) {
```
Will be better to rename "value" to "expectedLoginStatus".

emilshina commented on src/main/java/io/github/sskorol/assertions/ProductPageAssert.java:
```diff
@@ -1,29 +1,30 @@
 
-    public ProductPageAssert(final Product product) {
-        super(product, ProductPageAssert.class);
+    public ProductPageAssert(final ProductPage actual) {
```
The same errors as in LoginPageAssert class

lmalakhova committed:
Added the following:
- common refactoring according to requested changes.

sskorol commented on src/main/java/io/github/sskorol/assertions/LoginPageAssert.java:
```diff
@@ -12,19 +13,20 @@
 
-        if (!Objects.areEqual(loginStatus, value)) {
-            failWithMessage(ERROR_MESSAGE_TEMPLATE, "Status name", actual, loginStatus);
+        if (!Objects.areEqual(loginStatus, expectedLoginStatus)) {
```
you're comparing String with enum... extract enum value for further comparison

sskorol replied:
missed a getter call... seems good then
